frame/pablo/src/lib.rs

```diff
@@ -467,6 +541,139 @@ pub mod pallet {
+					current_timestamp,
+					rate_base,
+				)?,
+				compute_next_price_cumulative::<T::Convert, _, _>(
+					previous_price_cumulative.timestamp,
+					previous_price_cumulative.quote_price_cumulative,
+					current_timestamp,
+					rate_quote,
+				)?,
+			);
+			Ok((base_price_cumulative, quote_price_cumulative))
+		}
+
+		fn get_average_price(
+			current_base_price_cumulative: T::Balance,
+			previous_base_price_cumulative: T::Balance,
+			current_quote_price_cumulative: T::Balance,
+			previous_quote_price_cumulative: T::Balance,
+			elapsed: MomentOf<T>,
+		) -> Result<(Rate, Rate), DispatchError> {
+			let average_price_base = compute_price_average::<T::Convert, _, _>(
+				current_base_price_cumulative,
+				previous_base_price_cumulative,
+				elapsed,
+			)?;
+			let average_price_quote = compute_price_average::<T::Convert, _, _>(
+				current_quote_price_cumulative,
+				previous_quote_price_cumulative,
+				elapsed,
+			)?;
+			Ok((average_price_base, average_price_quote))
+		}
+
+		fn update_twap(pool_id: T::PoolId) -> Result<(), DispatchError> {
+			// update price cumulatives
+			let (base_price_cumulative, quote_price_cumulative) =
+				PriceCumulativeState::<T>::try_mutate(
+					pool_id,
+					|prev_price_cumulative| -> Result<(T::Balance, T::Balance), DispatchError> {
+						if let Some(previous_price_cumulative) = prev_price_cumulative {
+							let current_timestamp = T::Time::now();
+							let (base_price_cumulative, quote_price_cumulative) =
+								Self::get_next_price_cumulative(
+									pool_id,
+									previous_price_cumulative,
+								)?;
+							*prev_price_cumulative = Some(PriceCumulative {
+								timestamp: current_timestamp,
+								base_price_cumulative,
+								quote_price_cumulative,
+							});
+							Ok((base_price_cumulative, quote_price_cumulative))
+						} else {
+							Ok((T::Balance::zero(), T::Balance::zero()))
+						}
+					},
+				)?;
+			if base_price_cumulative != T::Balance::zero() &&
+				quote_price_cumulative != T::Balance::zero()
+			{
+				// update TWAP
+				TWAPState::<T>::try_mutate(
+					pool_id,
+					|prev_twap_state| -> Result<(), DispatchError> {
+						if let Some(twap_state) = prev_twap_state {
+							let previous_twap_state = twap_state.clone();
+							let current_timestamp = T::Time::now();
+							ensure!(
+								current_timestamp > previous_twap_state.timestamp,
+								DispatchError::Other("Stale TWAP Request Found")
+							);
+							let elapsed =
+								current_timestamp.saturating_sub(previous_twap_state.timestamp);
+							if elapsed >= T::TWAPInterval::get() {
+								let (average_price_base, average_price_quote) =
+									Self::get_average_price(
+										base_price_cumulative,
+										previous_twap_state.base_price_cumulative,
+										quote_price_cumulative,
+										previous_twap_state.quote_price_cumulative,
+										elapsed,
+									)?;
+								*prev_twap_state = Some(TimeWeightedAveragePrice {
```

Comment:
I wonder if this is the best way to keep twap updated as this is not accurate overtime as per the sampling rate/interval that it is supposed to follow. i.e without swap happening regularly the twap won't be calculated and updated in the storage on the expected sampling rate. Say the interval is 5 mins and the following sequence of updates
1st Min: cumulative calculated
2nd Min: Swap happens -> No twap update
6th Min: swap happens -> Twap updated -> **Not on time**
8th Min: Swap happens -> No Twap update
....
25th Min: Swap happens -> Twap updated -> **Not on time**

So there is a high probability of twap being outdated. Better way might be to build an extrinsic to get the TWAP at a given time by sampling the cumulative, which is always going to be up to date.

Reply:
Practically interval will be at-least as large as 1 hour. And given that real life pools will have large liquidity so even if TWAP is not updated as soon as interval passes , it will be still meaningful value.
@hussein-aitlahcen @KaiserKarel  any thoughts ?

Reply:
I think this should be fine, especially because we are going to enable it on pools that are "trusted", hopefully with some liquidity/swap happening within the interval. Otherwis we need to reverse back to my initial implementation using block hooks, i.e. updating the cumulative/TWAP in the hook when the interval passed while also updating the cumulative anytime when liquidity is moving.